### PR TITLE
Add error codes, allow running tests with missing features, stop hangs

### PR DESCRIPTION
**Pull request by Yasuke**

### CODE REVIEW

File: hs/src/Daml/Cucumber.hs [hunk 1]
```diff
@@ -79,39 +82,94 @@ runTestSuite (Opts folder mFeatureFile damlFolder) = do
...
+        Left error -> do
+          T.putStrLn $ "lsp session failed: " <> error
+          exitFailure
+
+        Right testResults -> do
+          success <- evaluateResults definedSteps testResults features
+          when (not $ Set.null missingSteps) $ do
+            putStrLn "Missing steps:"
+            for_ (Set.toList missingSteps) $ putStrLn . prettyPrintStep
+          when (not success) $ exitFailure
```
Comment (madeline-os) [inline hs/src/Daml/Cucumber.hs:104]: if we've allowed missing steps, does it still cause a failure at this point?

File: hs/src/Daml/Cucumber.hs [hunk 2]
```diff
@@ -79,39 +82,94 @@ runTestSuite (Opts folder mFeatureFile damlFolder) = do
...
-              _ -> putStrLn $ ("  " <> pretty <> " => DID NOT RUN")
-      pure ()
-  pure ()
+            _ -> do
+              modify (const False)
+              liftIO $ putStrLn $ ("  " <> pretty <> " => DID NOT RUN")
+
+  pure success
+
+  -- case Set.null missingSteps of
```
Comment (madeline-os) [inline hs/src/Daml/Cucumber.hs:140]: remove commented out code

File: hs/src/Daml/Cucumber.hs [hunk 3]
```diff
@@ -145,10 +203,10 @@ addFunction :: DamlFunc -> DamlScript -> DamlScript
 addFunction func state =
   state { damlFunctions = func : (damlFunctions state) }
 
-generateDamlSource :: Map Step StepFunc -> [Feature] -> State DamlScript ()
-generateDamlSource stepMapping features = do
+generateDamlSource :: Set Step -> Map Step StepFunc -> [Feature] -> State DamlScript ()
+generateDamlSource definedSteps stepMapping features = do
   for_ features $ \feature -> do
-    for_ (_feature_scenarios feature) $ \scenario -> do
+    for_ (filter (isScenarioRunnable definedSteps) $ _feature_scenarios feature) $ \scenario -> do
```
Comment (madeline-os) [inline hs/src/Daml/Cucumber.hs:209]: oh worm this will avoid running scenarios with undefined steps? got it

**Review by madeline-os — CHANGES_REQUESTED** [review]

File: hs/src/Daml/Cucumber.hs (hunk 1 above)
Comment (Yasuke) [reply hs/src/Daml/Cucumber.hs:104]: Nope, missing steps does not error if we allowed it, we still report them

**Review by madeline-os — APPROVED** [review]

**Pull request merged**